python/dgl/transform/module.py

```diff
@@ -0,0 +1,790 @@
+                                    return_counts=self.return_counts,
+                                    writeback_mapping=self.writeback_mapping,
+                                    copy_edata=True,
+                                    aggregator=self.aggregator)
+
+class LineGraph(BaseTransform):
+    r"""
+
+    Description
+    -----------
+    Return the line graph of the input graph.
+
+    The line graph :math:`L(G)` of a given graph :math:`G` is a graph where
+    the nodes in :math:`L(G)` correspond to the edges in :math:`G`. For a pair
+    of edges :math:`(u, v)` and :math:`(v, w)` in :math:`G`, there will be an
+    edge from the node corresponding to :math:`(u, v)` to the node corresponding to
+    :math:`(v, w)` in :math:`L(G)`.
+
+    This module only works for homogeneous graphs.
+
+    Parameters
+    ----------
+    backtracking : bool, optional
+        If True, the line graph node corresponding to :math:`(u, v)` will not have an
+        edge connecting to the line graph node corresponding to :math:`(v, u)`.
+
+    Example
+    -------
+
+    The following example uses PyTorch backend.
+
+    >>> import dgl
+    >>> import torch
+    >>> from dgl import LineGraph
+
+    Case1: Backtracking is True
+
+    >>> transform = LineGraph()
+    >>> g = dgl.graph(([0, 1, 1], [1, 0, 2]))
+    >>> g.ndata['h'] = torch.tensor([[0.], [1.], [2.]])
+    >>> g.edata['w'] = torch.tensor([[0.], [0.1], [0.2]])
+    >>> new_g = transform(g)
+    >>> print(new_g)
+    Graph(num_nodes=3, num_edges=3,
+          ndata_schemes={'w': Scheme(shape=(1,), dtype=torch.float32)}
+          edata_schemes={})
+    >>> print(new_g.edges())
+    (tensor([0, 0, 1]), tensor([1, 2, 0]))
+
+    Case2: Backtracking is False
+
+    >>> transform = LineGraph(backtracking=False)
+    >>> new_g = transform(g)
+    >>> print(new_g.edges())
+    (tensor([0]), tensor([2]))
+    """
+    def __init__(self, backtracking=True):
+        self.backtracking = backtracking
+
+    def __call__(self, g):
+        return functional.line_graph(g, backtracking=self.backtracking, shared=True)
+
+class KHopGraph(BaseTransform):
+    r"""
+
+    Description
+    -----------
+    Return the graph whose edges connect the :math:`k`-hop neighbors of the original graph.
+
+    This module only works for homogeneous graphs.
+
+    Parameters
+    ----------
+    k : int
+        The number of hops.
+
+    Example
+    -------
+
+    >>> import dgl
+    >>> from dgl import KHopGraph
+
+    >>> transform = KHopGraph(2)
+    >>> g = dgl.graph(([0, 1], [1, 2]))
+    >>> new_g = transform(g)
+    >>> print(new_g.edges())
+    (tensor([0]), tensor([2]))
+    """
+    def __init__(self, k):
+        self.k = k
+
+    def __call__(self, g):
+        return functional.khop_graph(g, self.k)
+
+class AddMetaPaths(BaseTransform):
+    r"""
+
+    Description
+    -----------
+    Add canonical edge types to an input graph based on given metapaths, as described in
```

Comment:
`Add new edges to an input graph based on given metapaths, ...`

Reply:
Changed.